### PR TITLE
Initial stub out of the CR repo API

### PR DESCRIPTION
**Pull request by JPercival**
Add an implementation of the Clinical Reasoning FHIR Repository to allow calling the available CR operations in hapi-fhir.

### CODE REVIEW

File: hapi-fhir-storage-cr/src/main/java/ca/uhn/fhir/cr/repo/RequestDetailsCloner.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+import ca.uhn.fhir.rest.api.server.SystemRequestDetails;
+
+/**
+ * This class produces partial clones of RequestDetails, the intent being to reuse the context of a
+ * RequestDetails object for reentrant calls. It retains header and tenancy information while
+ * scrapping everything else.
+ */
+class RequestDetailsCloner {
+
+	static DetailsBuilder startWith(RequestDetails theDetails) {
```
Comment (epeartree) [inline hapi-fhir-storage-cr/src/main/java/ca/uhn/fhir/cr/repo/RequestDetailsCloner.java:40]: same thing here with 'var'

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/api/server/SystemRequestDetails.java [hunk 2]
```diff
@@ -56,12 +56,22 @@ public class SystemRequestDetails extends RequestDetails {
 	 */
 	private RequestPartitionId myRequestPartitionId;
 
+	private IRestfulServerDefaults myServer;
+
 	public SystemRequestDetails() {
```
[the remark this answers is not in the document]
Comment (barhodes) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/api/server/SystemRequestDetails.java:63]: addressed

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/api/server/SystemRequestDetails.java [hunk 3]
```diff
@@ -56,12 +56,22 @@ public class SystemRequestDetails extends RequestDetails {
...
 		super(new MyInterceptorBroadcaster());
+		myServer = new MyRestfulServerDefaults();
 	}
 
 	public SystemRequestDetails(IInterceptorBroadcaster theInterceptorBroadcaster) {
 		super(theInterceptorBroadcaster);
+		myServer = new MyRestfulServerDefaults();
```
[the remark this answers is not in the document]
Comment (barhodes) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/api/server/SystemRequestDetails.java]: addressed

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/api/server/SystemRequestDetails.java [hunk 4]
```diff
@@ -56,12 +56,22 @@ public class SystemRequestDetails extends RequestDetails {
...
+	}
+
+	public SystemRequestDetails(RequestDetails theDetails) {
+		super(theDetails);
```
[the remark this answers is not in the document]
Comment (barhodes) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/api/server/SystemRequestDetails.java:72]: addressed

File: hapi-fhir-storage-cr/src/main/java/ca/uhn/fhir/cr/repo/HapiFhirRepository.java [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+public class HapiFhirRepository implements Repository {
+	private static final org.slf4j.Logger ourLog =
+			org.slf4j.LoggerFactory.getLogger(HapiFhirRepository.class);
+	private final DaoRegistry myDaoRegistry;
+	private final RequestDetails myRequestDetails;
+	private final RestfulServer myRestfulServer;
+
+	public HapiFhirRepository(DaoRegistry theDaoRegistry, RequestDetails theRequestDetails,
+			RestfulServer theRestfulServer) {
+		this.myDaoRegistry = theDaoRegistry;
```
[the remark this answers is not in the document]
Comment (barhodes) [reply hapi-fhir-storage-cr/src/main/java/ca/uhn/fhir/cr/repo/HapiFhirRepository.java]: addressed

File: hapi-fhir-storage-cr/src/main/java/ca/uhn/fhir/cr/repo/HapiFhirRepository.java [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+		this.myRequestDetails = theRequestDetails;
+		this.myRestfulServer = theRestfulServer;
+	}
+
+	@Override
+	public <T extends IBaseResource, I extends IIdType> T read(Class<T> theResourceType, I theId,
+			Map<String, String> theHeaders) {
+		var details = startWith(myRequestDetails).addHeaders(theHeaders).create();
```
[the remark this answers is not in the document]
Comment (barhodes) [reply hapi-fhir-storage-cr/src/main/java/ca/uhn/fhir/cr/repo/HapiFhirRepository.java:77]: Yes, I have confirmation from JP that var is acceptable

File: hapi-fhir-storage-cr/src/main/java/ca/uhn/fhir/cr/repo/HapiFhirRepository.java [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+				theRequestDetails, myRestfulServer.getDefaultResponseEncoding());
+		var linkEncoding = theRequestDetails.getParameters().containsKey(Constants.PARAM_FORMAT)
+				&& responseEncoding != null ? responseEncoding.getEncoding() : null;
+
+		return (B) BundleProviderUtil.createBundleFromBundleProvider(myRestfulServer,
+				theRequestDetails, count, linkSelf, includes, theBundleProvider, start, bundleType,
+				linkEncoding, thePagingAction);
+	}
+
+	// TODO: The main use case for this is paging through Bundles, but I suppose that technically
```
[the remark this answers is not in the document]
Comment (barhodes) [reply hapi-fhir-storage-cr/src/main/java/ca/uhn/fhir/cr/repo/HapiFhirRepository.java:173]: Yes, there is still work to be done here and this comment reflects that

File: hapi-fhir-storage-cr/src/main/java/ca/uhn/fhir/cr/repo/RequestDetailsCloner.java (hunk 1 above)
Comment (barhodes) [reply hapi-fhir-storage-cr/src/main/java/ca/uhn/fhir/cr/repo/RequestDetailsCloner.java:40]: addressed

**Pull request merged**